### DISCUSSION TITLE
Добавление комментария из временной таблицы

### DISCUSSION THREAD

**Discussion by juramebel** [totumonline/totum-ru-issues-and-discussions#40] [Как это сделать?] [answered]
Здравтвуйте!
Есть таблица циклов техкарт. В каждой - производственные операции. 
У каждой операции есть поле - комментарии. 
Оно нужно ,чтобы например фиксировать причины приостановки.
У рабочих в цеху есть отдельный интерфейс - временная таблица. Только с текущими операциями из этих циклов.
Стоит задача, чтобы он видел в своей временной таблице эти колмментарии и мог добавить новый, не заходя в саму техкарту (внутрь расчетной таблицы циклов).
Подскажите, пожалуйста, как сделать.

![image](https://user-images.githubusercontent.com/19141501/143386493-8379ea45-6d8f-4a97-8cbf-51170f7f0b04.png)

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Поле Комментарий имеет встроенный функционал в виде отслеживания прочитанных сообщений — поэтому мы отключили ему код. Тк кода нет, то нельзя просто взять и сделать поле Комментарий во временной и через select забрать значения из таблицы в цикле. Но можно это эмулировать:

![Снимок экрана 2021-11-25 в 15 22 28](https://user-images.githubusercontent.com/55755565/143441319-b27479ad-6840-4917-ab8f-363e75381c35.png)

Что мы делаем логически:

1. Добавляем во временную поле типа Строка с заблокированным изменением.

2. Делаем форматирование которое:

   + Берет значение из поля Комментарий внутри цикла
   
   + Обрабатывает этот список делая нормальный формат датам и подставляя ФИО пользователей

   + Выводим последнее сообщение в поле через setFormat параметр text

   + Выводим все сообщения в контекстную панель
   
   + Делаем кнопку в контекстной панели, которая:
   
     + Вызывает окно ввода
     
     + Добавляет комментарий через set внутрь цикла
     
     + Обновляет временную таблицу — refresh: "recalculate"

Вот как выглядит код форматирования (его прям целиком можно использовать, сделав замены name таблиц и полей):

```

f1=: setFormat(text: str`"Last:" ++ $replace[$count]`)

	count: listCount(list: $sel) - 1

p1=: panelHtml(html: str`$join + "<br><br>"`)

	join: listJoin(list: $replace; str: "<br><br>")

replace: listReplace(list: $sel; action: $val; value: "value")

		~sel: select(table: 's1_in_cycle'; cycle: #cycle; field: 'comments'; where: 'id' = #insert[id])
	
	val: str`$df ++ "||" ++ $user + ":" ++ $#value[2]`
			df: dateFormat(date: $#value[0]; format: "d.m")
			user: select(table: 'users'; field: 'fio'; where: 'id' = $#value[1])
	
p2=: panelButtons(button: $button)
	
	button: rowCreate(field: "text" = "ADD COMMENT"; field: "code" = $linktoinput; field: "vars" = $vars)
	
		linktoinput: "=: linkToInput(title: 'Add comment'; code: $#code; var: 'cycle' = #cycle; var: 'insertid' = #insert[id]; refresh: 'recalculate'; button: 'Add')"
		vars: rowCreate(field: "code" = $code)
			code: "=: set(table: 's1_in_cycle'; cycle: $#cycle; field: 'comments' = $#input; where: 'id' = $#insertid)"

		

```

**Reply by juramebel** [reply]
Спасибо, в целом работает!
Но почему задваивает и появляются строки в квадратных скобках
![Снимок 2](https://user-images.githubusercontent.com/19141501/143456061-0715d2c3-52f6-46bf-9ed3-1a732b3a17cf.PNG)
![Снимок](https://user-images.githubusercontent.com/19141501/143456071-670d7b89-7a13-4b6e-b3bb-a020cfe0f198.PNG)

**Reply by ttm-alexeypolunin** [reply]
Там сейчас стоит тип поля Данные и выполняется код, который берет значение в него. 

Они и показываются. В 2 будет параметр `textasvalue` для `setFormat`, который позволит заменять для пользователя значение на `text`.

**Сейчас надо отключить Код — что бы значения в этом поле были пустые. Все формируется в Форматировании**

**Reply by juramebel** [reply]
Код отключен. Это я понял про форматирование.

**Reply by juramebel** [reply]
В общем что то странное. Я выключил и код и форматирование, несколько раз все пересчитал, а значения эти все равно висят) Код то нормально работает

**Reply by juramebel** [reply]
Решил только так - сделал доступным ручное редактирование, удалил все содержимое, потом опять все заблокировал.

**Reply by ttm-alexeypolunin** [reply]
Это точно во временной таблице сделано, или внутри цикла? Если во временной, то переход в нее из Дерева приводит к созданию ее нового варианта.

Описанное поведение, когда был код, а потом его отключили, а значения остались — это нормально для простых и расчетных таблиц — значение то уже посчитано, оно и остается (и для временных, пока вы смотрите на один из ее клонов). Это надо учитывать, когда отключаются коды.

**Reply by juramebel** [reply]
А, верно.
Я просто делал такое поле и в таблице циклов и во временной. 
Такое поведение было в таблице циклов. 
Спасибо, буду знать.
Теперь все работает отлично!)
